### PR TITLE
Find and execute plugins

### PR DESCRIPTION
**Pull request by tealeg**
This PR makes Pink discover plugins, and, if required, execute them via their manifest.json definiton. 

It covers:
- Listing plugins installed
- triggering help
- causing invocation of plugins in line with their manlfest.json files.

### CODE REVIEW

File: plugins_test.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	tearDown, pluginRoot := setUp(t)
+	defer tearDown()
+	assertPlugins := func(t *testing.T, root string, expected []string) {
+		t.Run(root, func(t *testing.T) {
+			plugins, err := findPlugins(root)
+			if err != nil {
+				t.Fatalf(err.Error())
+			}
+			require.ElementsMatch(t, expected, plugins)
+
```
Comment (YassineE) [inline plugins_test.go]:
```suggestion
```

File: plugins_test.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	assertPlugins(t, pluginRoot, []string{"system", "foo"})
+	assertPlugins(t, filepath.Join(pluginRoot, "system"), []string{"run", "stop"})
+	assertPlugins(t, filepath.Join(pluginRoot, "system", "run"), []string{})
+	assertPlugins(t, filepath.Join(pluginRoot, "system", "stop"), []string{})
+	assertPlugins(t, filepath.Join(pluginRoot, "foo"), []string{})
+}
+
+// IsInvokable indicates the leaves of the command tree
```
Comment (YassineE) [inline plugins_test.go]:
```suggestion
// IsInvokable indicates the leaves of the command tree.
```

File: plugins_test.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+	require.False(t, inv)
+	inv, err = isInvokable(filepath.Join(pluginRoot, "system", "run"))
+	require.True(t, inv)
+	inv, err = isInvokable(filepath.Join(pluginRoot, "system", "stop"))
+	require.True(t, inv)
+	inv, err = isInvokable(filepath.Join(pluginRoot, "system", "run", "up"))
+	require.False(t, inv)
+}
+
+// DispatchCommand finds the appropriate plugins and passes it the right portion of the arguments
```
Comment (YassineE) [inline plugins_test.go]:
```suggestion
// DispatchCommand finds the appropriate plugins and passes it the right portion of the arguments.
```

**Review by yaziine — APPROVED** [review]
Few comments but otherwise it LGTM 👍

File: command.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"context"
+	"fmt"
+)
+
+// Help will list the plugins installed below a given root
+func Help(root string) {
+	plugins, err := findPlugins(root)
+	if err != nil {
+		panic(err.Error())
```
[the remark this answers is not in the document]
Comment (tealeg) [reply command.go]: Essentially, yes.  It's something very fundamental that's wrong - we could probably do something better, but it wasn't a priority at this early stage.

File: cmd/pink/main.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	"log"
+	"os"
+	"path/filepath"
+
+	"github.com/heetch/pink"
+)
+
+// getPinkDir returns the default state directory for pink
+func getPinkDir() string {
+	return filepath.Join(os.Getenv("HOME"), ".pink")
```
[the remark this answers is not in the document]
Comment (tealeg) [reply cmd/pink/main.go]: That would be hard to do.  File operations will give meaningful errors if we look for non-existant paths, so this shouldn't really be an issue.  Besides, if people are fiddling with their `$HOME` they're going to have bigger problems than this ;-)

File: cmd/pink/main.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+func main() {
+	pd := getPinkDir()
+	if err := ensurePinkDirs(pd); err != nil {
+		log.Fatalf(err.Error())
+	}
+	plugD := pink.GetPluginsDir(pd)
+	// If we don't pass any arguments default to running help on
+	// the root of the plugins directory
```
[the remark this answers is not in the document]
Comment (tealeg) [reply cmd/pink/main.go]: It's missing a comma after the word "arguments".

[the remark this answers is not in the document]
Comment (tealeg) [reply cmd/pink/main.go]: Fixed

File: command.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+	if len(plugins) == 0 {
+		fmt.Println("No pink plug-ins are currently installed")
+		return
+	}
+	fmt.Println("The following pink plugins are installed:")
+	for _, p := range plugins {
+		fmt.Printf("\t%s\n", p)
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (tealeg) [reply command.go]: Yup.  Very good call.  Done.

File: manifest.go [hunk 8]
```diff
@@ -59,3 +61,23 @@ func validateManifest(m *Manifest) error {
 
 	return nil
 }
+
+//
```
[the remark this answers is not in the document]
Comment (tealeg) [reply manifest.go]: Yes, Done.

File: manifest.go [hunk 9]
```diff
@@ -59,3 +61,23 @@ func validateManifest(m *Manifest) error {
...
+			PluginDir: pluginDir,
+		}
+	case "docker":
+		client, err := client.NewEnvClient()
+		if err != nil {
+			return err
+		}
+		inv = NewDockerInvoker(client, os.Stdout, os.Stderr)
+	default:
+		return errors.Errorf("unsupported invoker '%s', only 'executable' is currently supported", m.Invoker)
```
[the remark this answers is not in the document]
Comment (tealeg) [reply manifest.go]: Indeed.

File: cmd/pink/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (YassineE) [reply cmd/pink/main.go]: Indeed 😄

**Pull request merged**